code/game/machinery/spaceheater.dm

```diff
@@ -322,37 +317,27 @@
 	QDEL_NULL(beaker)
 
 /obj/machinery/space_heater/improvised_chem_heater/process(seconds_per_tick)
-	if(!on)
-		update_appearance()
-		return PROCESS_KILL
-
-	if(!is_operational || !cell || cell.charge <= 0)
-		on = FALSE
-		update_appearance()
+	if(!on || !is_operational || QDELETED(cell) || cell.charge <= 1 || QDELETED(beaker))
+		if (on) // If it's broken, turn it off too
+			on = FALSE
+			update_appearance()
 		return PROCESS_KILL
 
-	if(!beaker)//No beaker to heat
-		update_appearance()
-		return
-
 	if(beaker.reagents.total_volume)
 		var/power_mod = 0.1 * chem_heating_power
 		switch(set_mode)
 			if(HEATER_MODE_AUTO)
 				power_mod *= 0.5
-				beaker.reagents.adjust_thermal_energy((target_temperature - beaker.reagents.chem_temp) * power_mod * seconds_per_tick * SPECIFIC_HEAT_DEFAULT * beaker.reagents.total_volume)
-				beaker.reagents.handle_reactions()
 			if(HEATER_MODE_HEAT)
 				if(target_temperature < beaker.reagents.chem_temp)
 					return
-				beaker.reagents.adjust_thermal_energy((target_temperature - beaker.reagents.chem_temp) * power_mod * seconds_per_tick * SPECIFIC_HEAT_DEFAULT * beaker.reagents.total_volume)
 			if(HEATER_MODE_COOL)
 				if(target_temperature > beaker.reagents.chem_temp)
 					return
-				beaker.reagents.adjust_thermal_energy((target_temperature - beaker.reagents.chem_temp) * power_mod * seconds_per_tick * SPECIFIC_HEAT_DEFAULT * beaker.reagents.total_volume)
-		var/required_energy = heating_power * seconds_per_tick * (power_mod * 4)
-		cell.use(required_energy / efficiency)
+
+		beaker.reagents.adjust_thermal_energy((target_temperature - beaker.reagents.chem_temp) * power_mod * seconds_per_tick * beaker.reagents.heat_capacity() * beaker.reagents.total_volume)
```

Comment:
Don't scale by the reagent volume, because heat_capacity() gets the heat capacity and not the specific heat.

Also make sure the calculated energy cost of the cell won't be above the charge of the cell. Basically just a min check on the absolute of the thermal energy change and the cell charge scaled inversely by the heating conversion, then scale that by the sign of the thermal energy change.